sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/CachedBatchColumnVector.java

```diff
@@ -0,0 +1,421 @@
+ */
+package org.apache.spark.sql.execution.vectorized;
+
+import java.nio.ByteBuffer;
+
+import org.apache.spark.memory.MemoryMode;
+import org.apache.spark.sql.execution.columnar.*;
+import org.apache.spark.sql.types.*;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * A column backed by an in memory JVM array.
+ */
+public final class CachedBatchColumnVector extends ColumnVector implements java.io.Serializable {
+
+  // keep compressed data
+  private byte[] buffer;
+
+  // whether a row is already extracted or not. If extractTo() is called, set true
+  // e.g. when isNullAt() and getInt() ara called, extractTo() must be called only once
+  private boolean[] calledExtractTo;
+
+  // accessor for a column
+  private transient ColumnAccessor columnAccessor;
+
+  // a row where the compressed data is extracted
+  private transient ColumnVector columnVector;
+
+  // an accessor uses only row 0 in columnVector
+  private final int ROWID = 0;
+
+
+  public CachedBatchColumnVector(byte[] buffer, int numRows, DataType type) {
+    super(numRows, DataTypes.NullType, MemoryMode.ON_HEAP);
+    initialize(buffer, type);
+    reserveInternal(numRows);
+    reset();
+  }
+
+  @Override
+  public long valuesNativeAddress() {
+    throw new RuntimeException("Cannot get native address for on heap column");
+  }
+  @Override
+  public long nullsNativeAddress() {
+    throw new RuntimeException("Cannot get native address for on heap column");
+  }
+
+  @Override
+  public void close() {
+  }
+
+  private void setColumnAccessor() {
+    ByteBuffer byteBuffer = ByteBuffer.wrap(buffer);
+    columnAccessor = ColumnAccessor$.MODULE$.apply(type, byteBuffer);
+    calledExtractTo = new boolean[capacity];
+  }
+
+  // call extractTo() before getting actual data
+  private void prepareAccess(int rowId) {
+    if (!calledExtractTo[rowId]) {
+      assert (columnAccessor.hasNext());
+      columnAccessor.extractTo(columnVector, ROWID);
+      calledExtractTo[rowId] = true;
+    }
+  }
+
+  //
+  // APIs dealing with nulls
+  //
+
+  @Override
+  public void putNotNull(int rowId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putNull(int rowId) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putNulls(int rowId, int count) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putNotNulls(int rowId, int count) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean isNullAt(int rowId) {
+    prepareAccess(rowId);
+    return columnVector.isNullAt(ROWID);
+  }
+
+  //
+  // APIs dealing with Booleans
+  //
+
+  @Override
+  public void putBoolean(int rowId, boolean value) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putBooleans(int rowId, int count, boolean value) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean getBoolean(int rowId) {
```

Comment:
We do not support reading values in a random order. This is because implementation of `CompressionScheme` (e.g. `IntDelta`) supports only sequential access.